include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+       * computations. The operations that are performed with it are recorded on
+       * the tape and will be considered in the computation of dependent
+       * variable values.
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      mark_independent_variable(ad_type &out, const unsigned int index) const;
+
+      /**
+       * Finalize the state of the independent variables before use.
+       *
+       * This step and the storage of the independent variables is done
+       * separately because some derived classes may offer the capability
+       * to add independent variables in a staggered manner. This function
+       * is to be triggered when these values are considered finalized
+       * and we can safely initialize the sensitive equivalents of those
+       * values.
+       */
+      void
+      finalize_sensitive_independent_variables() const;
+
+      /**
+       * Initialize an independent variable $X_{i}$.
+       *
+       * @param[out] out An auto-differentiable number that is ready for use in
+       * standard computations. The operations that are performed with it are
+       * not recorded on the tape, and so should only be used when not in
+       * recording mode.
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      initialize_non_sensitive_independent_variable(
+        ad_type &          out,
+        const unsigned int index) const;
+
+      /**
+       * The number of independent variables that have been manipulated within a
+       * set of operations.
+       */
+      unsigned int
+      n_touched_independent_variables() const;
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * The set of dependent variables $\mathbf{f}(\mathbf{X})$ of which the
+       * derivatives with respect to $\mathbf{X}$ will be computed.
+       *
+       * The gradients and Hessians of these dependent variables will be
+       * computed at the values $\mathbf{X}$ set with the set_sensitivity_value
+       * function.
+       *
+       * @note These are stored as an @p ad_type so that we can use them to
+       * compute function values and directional derivatives in the case that
+       * tapeless numbers are used
+       */
+      std::vector<ad_type> dependent_variables;
+
+      /**
+       * A list of dependent variables.
+       */
+      std::vector<bool> touched_dependent_variables;
+
+      /**
+       * Reset the boolean vector that indicates which independent variables
+       * that have been manipulated by the current set of operations.
+       * All entries in the vector are set to the value of the @p flag.
+       */
+      void
+      reset_touched_dependent_variables(const bool flag = false);
+
+      /**
+       * The number of dependent variables that have been registered.
+       */
+      unsigned int
+      n_touched_dependent_variables() const;
```

Comment:
I get confused between the terms "touched" and "registered". Is there a difference? If not, maybe settle on one of the terms?